Make `(try_)subst_and_normalize_erasing_regions` take `EarlyBinder`

kylematsuda opened the pull request:
Changes `subst_and_normalize_erasing_regions` and `try_subst_and_normalize_erasing_regions` to take  `EarlyBinder<T>` instead of `T`. 

(related to #105779)

This was suggested by @BoxyUwU in https://github.com/rust-lang/rust/pull/107753#discussion_r1105828139. After changing `type_of` to return `EarlyBinder`, there were several places where the binder was immediately skipped to call `tcx.subst_and_normalize_erasing_regions`, only for the binder to be reconstructed inside of that method.

r? @BoxyUwU

kylematsuda committed:
make (try_)subst_and_normalize_erasing_regions take EarlyBinder

jackh726 commented on src/tools/clippy/clippy_lints/src/methods/unnecessary_to_owned.rs:
```diff
@@ -435,7 +435,7 @@ fn can_change_type<'a>(cx: &LateContext<'a>, mut expr: &'a Expr<'a>, mut ty: Ty<
                         if output_ty.contains(*param_ty) {
                             if let Ok(new_ty)  = cx.tcx.try_subst_and_normalize_erasing_regions(
-                                new_subst, cx.param_env, output_ty) {
+                                new_subst, cx.param_env, EarlyBinder(output_ty)) {
```
So, seems like the `subst_identity().skip_binder()` call to create this isn't correct...

kylematsuda replied:
Hmm, yeah I agree that this looks weird now. Would it be better as `skip_binder().skip_binder()`? Or are you envisioning a larger change here?

jackh726 replied:
I have to look a bit more at this. But likely the easiest thing is to either just leave it as-is with a comment, or map/skip_binder above.

kylematsuda replied:
Ah okay thanks! I changed it to `skip_binder()` above in the most recent commit, but happy to try to improve it further.

Yeah, I also thought about mapping it, it seems like we always skip the inner `Binder` there but want to keep the `EarlyBinder` around, so could do something like

```rust
let fn_sig = cx.tcx.fn_sig(def_id).map_bound(|t| t.skip_binder());
```

(and then replace the calls to  `.inputs()` with `.skip_binder().inputs()`)

but I wasn't sure if this was more or less readable than how it is now

BoxyUwU replied:
I think this PR doesn't necessarily need to fix this, at minimum we should probably open a clippy issue about it and add a FIXME before this PR lands

jackh726 commented on compiler/rustc_middle/src/ty/instance.rs:
```diff
@@ -657,7 +657,7 @@ impl<'tcx> Instance<'tcx> {
     {
         if let Some(substs) = self.substs_for_mir_body() {
-            tcx.try_subst_and_normalize_erasing_regions(substs, param_env, v)
+            tcx.try_subst_and_normalize_erasing_regions(substs, param_env, ty::EarlyBinder(v))
```
Should these functions also take an `EarlyBinder`?

kylematsuda replied:
Hmm yeah it seems like they probably should. I tried this in a new commit, please let me know what you think.

Something I'm generally wondering about is how far to "bubble up" these `EarlyBinder` changes, like `subst_mir_and_normalize_erasing_regions` doesn't get called too much directly but it's wrapped by some other methods that get called more, and I just left those as `T` instead of changing them to take `EarlyBinder<T>`. Happy to start changing those as well, just wanted to ask first

jackh726 replied:
I can look later today. In the most ideal world, we bubble up EarlyBinder creation to where we create the params and ReEarlyBound. For this PR, it'll likely be some balance of "makes sense" and "small enough"

kylematsuda committed:
use EarlyBinder in tcx.(try_)subst_mir_and_normalize_erasing_regions

kylematsuda committed:
make subst_mir take EarlyBinder

kylematsuda commented on compiler/rustc_middle/src/ty/instance.rs:
```diff
@@ -618,14 +618,15 @@ impl<'tcx> Instance<'tcx> {
+            v.subst(tcx, substs)
         } else {
-            *v
+            v.subst_identity()
```
Should this be `no_bound_vars().unwrap()` instead? The doc comment on `substs_for_mir_body` above seemed to suggest it might but I wasn't sure

BoxyUwU replied:
Yeah I think so

kylematsuda replied:
So I tried `no_bound_vars().unwrap()` here and this ends up panicking when building `core` :grimacing: 

Specifically:

```plaintext
query stack during panic:
#0 [optimized_mir] optimizing MIR for `mem::maybe_uninit::<impl at library/core/src/mem/maybe_uninit.rs:270:1: 270:23>::write_slice_cloned::<impl at library/core/src/mem/maybe_uninit.rs:1128:9: 1128:42>::drop`
end of query stack
```

Re-reading the doc comment on `substs_for_mir_body`, it says (paraphrasing) "When this returns `None`, we don't want to do a substitution, since it has effectively already been done." I suppose that doesn't actually imply that there are no bound vars?

Maybe this should actually be `skip_binder`? I changed it to this in the latest commit (and in `(try_)subst_mir_and_normalize_erasing_regions`, which have a similar thing happening)

bors commented:
:umbrella: The latest upstream changes (presumably #96840) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

BoxyUwU commented:
I will r+ this once its been rebased again and that `subst_identity()` has been changed to `no_bound_vars().unwrap()`.  sorry for taking so long to look over this, thanks for doing this

kylematsuda committed:
changes from review: add FIXME to clippy and change subst_identity to skip_binder in mir subst methods

kylematsuda commented:
Hi @BoxyUwU, thanks for taking a look! Sorry, I've also been busy recently and sort of forgot about this PR :sweat_smile:

I rebased and added the FIXME as you suggested, the `no_bound_vars().unwrap()` doesn't seem to work though unfortunately (see comment above)

BoxyUwU commented:
@bors r+

bors commented:
:pushpin: Commit d27f40175fb7221bc5fa4c19ff75e211947222e8 has been approved by `BoxyUwU`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=BoxyUwU d27f40175fb7221bc5fa4c19ff75e211947222e8 -->
<!-- homu: {"type":"Approved","sha":"d27f40175fb7221bc5fa4c19ff75e211947222e8","approver":"BoxyUwU","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
